### PR TITLE
Debug dy port breakout fixes sonic_utils.

### PR DESCRIPTION
**Pull request by praveen-li**
**- What I did**
[main.py]: Fixing issues found during Dy Port Break Testing.
Added more tests in config_mgmt.
Sorting Data before writing it to Config DB.

**- How I did it**
For Add Ports: Python Click was providing Delete Port List. Corrected it.
PORT Table should be passed from Python Click instead of just Ports dictionary. Fixed it.
Added more tests in config_mgmt.
Sorting Data before writing it to Config DB.

**- How to verify it**
```
Running Breakout Mode : 4x25G[10G] 
Target Breakout Mode : 1x100G[40G]

Ports to be deleted : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}
Ports to be added : 
 {
    "Ethernet0": "100000"
}

After running Logic to limit the impact

Final list of ports to be deleted : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
} 
Final list of ports to be added :  
 {
    "Ethernet0": "100000"
}
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet2
Find dependecies for port Ethernet3
Find dependecies for port Ethernet0
Find dependecies for port Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been deleted successfully ***


Start Port Addition
Merge Port Config for ['Ethernet0']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been added successfully ***

Breakout process got successfully completed.
ps

                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       ce0( 68)  down   4  100G  FD   SW  No   Forward          None    D    KR4  9412    No      
       xe0( 69)  !ena   1     -       SW  No   Forward          None    D   None  9122    No      
       xe1( 70)  !ena   2     -       SW  No   Forward          None    D   None  9122    No      
       xe2( 71)  !ena   1     -       SW  No   Forward          None    D   None  9122    No      
       xe3( 72)  !ena   1   25G  FD   SW  No   Forward          None   FA     KR  9122    No      

Running Breakout Mode : 1x100G[40G] 
Target Breakout Mode : 2x50G

Ports to be deleted : 
 {
    "Ethernet0": "100000"
}
Ports to be added : 
 {
    "Ethernet2": "50000", 
    "Ethernet0": "50000"
}

After running Logic to limit the impact

Final list of ports to be deleted : 
 {
    "Ethernet0": "100000"
} 
Final list of ports to be added :  
 {
    "Ethernet2": "50000", 
    "Ethernet0": "50000"
}
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Deleting Port: Ethernet0
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been deleted successfully ***


Start Port Addition
Merge Port Config for ['Ethernet2', 'Ethernet0']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been added successfully ***

Breakout process got successfully completed.
ps

                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       xe0( 68)  down   2   50G  FD   SW  No   Forward          None    D    KR2  9412    No      
       xe1( 69)  !ena   1     -       SW  No   Forward          None    D   None  9122    No      
       xe2( 70)  down   2   50G  FD   SW  No   Forward          None    D    KR2  9412    No      
       xe3( 71)  !ena   1     -       SW  No   Forward          None    D   None  9122    No      

Running Breakout Mode : 2x50G 
Target Breakout Mode : 4x25G[10G]

Ports to be deleted : 
 {
    "Ethernet2": "50000", 
    "Ethernet0": "50000"
}
Ports to be added : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}

After running Logic to limit the impact

Final list of ports to be deleted : 
 {
    "Ethernet2": "50000", 
    "Ethernet0": "50000"
} 
Final list of ports to be added :  
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet2
Find dependecies for port Ethernet0
Deleting Port: Ethernet2
Deleting Port: Ethernet0
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been deleted successfully ***


Start Port Addition
Merge Port Config for ['Ethernet2', 'Ethernet3', 'Ethernet0', 'Ethernet1']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been added successfully ***

Breakout process got successfully completed.
ps

                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       xe0( 68)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe1( 69)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe2( 70)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe3( 71)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      

Running Breakout Mode : 4x25G[10G] 
Target Breakout Mode : 1x100G[40G]

Ports to be deleted : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}
Ports to be added : 
 {
    "Ethernet0": "100000"
}

After running Logic to limit the impact

Final list of ports to be deleted : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
} 
Final list of ports to be added :  
 {
    "Ethernet0": "100000"
}
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet2
Find dependecies for port Ethernet3
Find dependecies for port Ethernet0
Find dependecies for port Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been deleted successfully ***


Start Port Addition
Merge Port Config for ['Ethernet0']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been added successfully ***

Breakout process got successfully completed.
ps

                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       ce0( 68)  down   4  100G  FD   SW  No   Forward          None    D    KR4  9412    No      
       xe0( 69)  !ena   1     -       SW  No   Forward          None    D   None  9412    No      
       xe1( 70)  !ena   2     -       SW  No   Forward          None    D   None  9412    No      
       xe2( 71)  !ena   1     -       SW  No   Forward          None    D   None  9412    No      
       xe3( 72)  !ena   1   25G  FD   SW  No   Forward          None   FA     KR  9122    No      

Running Breakout Mode : 1x100G[40G] 
Target Breakout Mode : 4x25G[10G]

Ports to be deleted : 
 {
    "Ethernet0": "100000"
}
Ports to be added : 
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}

After running Logic to limit the impact

Final list of ports to be deleted : 
 {
    "Ethernet0": "100000"
} 
Final list of ports to be added :  
 {
    "Ethernet2": "25000", 
    "Ethernet3": "25000", 
    "Ethernet0": "25000", 
    "Ethernet1": "25000"
}
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Deleting Port: Ethernet0
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been deleted successfully ***


Start Port Addition
Merge Port Config for ['Ethernet2', 'Ethernet3', 'Ethernet0', 'Ethernet1']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

*** Ports have been added successfully ***

Breakout process got successfully completed.
ps

                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       xe0( 68)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe1( 69)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe2( 70)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
       xe3( 71)  down   1   25G  FD   SW  No   Forward          None    D     KR  9412    No      
```

Test Run After Commit 2:
```
admin@lnos-x1-a-asw01:~/pc_testing$ sudo python config_mgmt.py  -n 4x25G[10G] -c 1x100G[40G] -l True
Test Run Delete Ports
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Deleting Port: Ethernet0
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB
Verify Port Table in config DB for Deletion

***Port Deletion Test Passed***

Wait for 5 Secs

Start Port Addition
Generating default config for ['Ethernet0', 'Ethernet1', 'Ethernet2', 'Ethernet3']
Merge Port Config for ['Ethernet0', 'Ethernet1', 'Ethernet2', 'Ethernet3']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB
Verify Port Table in config DB for Addition

***Port Addition Test Passed***

admin@lnos-x1-a-asw01:~/pc_testing$ sudo python config_mgmt.py  -c 4x25G[10G] -n 2x50G -l True
Test Run Delete Ports
Loaded below Yang Models
['sonic-acl', 'sonic-head', 'sonic-interface', 'sonic-port', 'sonic-portchannel', 'sonic-vlan']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Find dependecies for port Ethernet1
Find dependecies for port Ethernet2
Find dependecies for port Ethernet3
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB
Verify Port Table in config DB for Deletion

***Port Deletion Test Passed***

Wait for 5 Secs

Start Port Addition
Generating default config for ['Ethernet0', 'Ethernet2']
Merge Port Config for ['Ethernet0', 'Ethernet2']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB
Verify Port Table in config DB for Addition

***Port Addition Test Passed***
```

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

File: config/config_mgmt.py [hunk 1]
```diff
@@ -77,7 +76,7 @@ def logInFile(self, header="", obj=None, json=False):
                 df.write('\n\n{}: {}\n'.format(time, header))
                 if json:
                     dump(obj, df, indent=4)
-                else:
+                elif obj:
                     #print(obj)
                     df.write('{}: {}'.format(time, obj))
```
Comment (li-pingmao) [inline config/config_mgmt.py:81]: else:  statement to handle other cases?

File: config/config_mgmt.py [hunk 2]
```diff
@@ -540,8 +545,9 @@ def testRun_Config_Reload_load():
 
     return
 
-def testRun_Delete_Add_Ports():
+def testRun_Delete_Add_Ports_1():
```
Comment (li-pingmao) [inline config/config_mgmt.py]: Add descriptions for 3 test functions?    for example,  Break out from 100G -> 4x25G etc.

File: config/config_mgmt.py [hunk 3]
```diff
@@ -555,27 +561,30 @@ def testRun_Delete_Add_Ports():
...
+                    "lanes": "67,68",
                     "description": "",
                     "speed": "50000"
             }
         }
     }
 
-    ret = cm.addPorts(ports=['Ethernet0', 'Ethernet2'], portJson=portJson)
+    #ret = cm.addPorts(ports=['Ethernet0', 'Ethernet2'], portJson=portJson, loadDefConfig=False)
+    ret = cm.addPorts(ports=['Ethernet0', 'Ethernet2'], portJson=portJson, loadDefConfig=True)
```
Comment (li-pingmao) [inline config/config_mgmt.py]: loadDefConfig = False case  will be tested too?

File: config/config_mgmt.py [hunk 4]
```diff
@@ -584,7 +593,125 @@ def testRun_Delete_Add_Ports():
...
+                    "admin_status": "up",
+                    "lanes": "65,66,67,68",
+                    "description": "",
+                    "speed": "100000"
+            }
+        }
+    }
+
+    #ret = cm.addPorts(ports=['Ethernet0'], portJson=portJson, loadDefConfig=False)
+    ret = cm.addPorts(ports=['Ethernet0'], portJson=portJson, loadDefConfig=True)
```
Comment (li-pingmao) [inline config/config_mgmt.py]: configuration is defined in portJson,   should the loadDefConfig be False?

File: config/config_mgmt.py [hunk 5]
```diff
@@ -584,7 +593,125 @@ def testRun_Delete_Add_Ports():
...
+                "description": "",
+                "index": "0",
+                "lanes": "68",
+                "speed": "25000"
+            }
+        }
+    }
+
+    #ret = cm.addPorts(ports=['Ethernet0', 'Ethernet1', 'Ethernet2', 'Ethernet3'], portJson=portJson, loadDefConfig=False)
+    ret = cm.addPorts(ports=['Ethernet0', 'Ethernet1', 'Ethernet2', 'Ethernet3'], portJson=portJson, loadDefConfig=True)
```
Comment (li-pingmao) [inline config/config_mgmt.py]: If configuration is defined in portJson,  should it be loadDefConfig=False?

File: config/config_mgmt.py [hunk 6]
```diff
@@ -584,7 +593,125 @@ def testRun_Delete_Add_Ports():
...
+    if len(argv) > 1:
+        num = argv[1]
+    else:
+        print("Please provide a number 1-3")
+        return
+
     #testRun_Config_Reload_load()
-    testRun_Delete_Add_Ports()
-    #test_sonic_yang()
+    if num == '1':
```
Comment (li-pingmao) [inline config/config_mgmt.py]: Can you add short description of options "1", "2" and "3" for different breakout modes to be tested?

File: config/main.py [hunk 7]
```diff
@@ -1444,132 +1551,116 @@ def speed(ctx, interface_name, interface_speed, verbose):
         command += " -vv"
     run_command(command, display_cmd=verbose)
 
+
+
+
```
Comment (li-pingmao) [inline config/main.py]: Remove extra lines.

File: config/main.py [hunk 8]
```diff
@@ -1444,132 +1551,116 @@ def speed(ctx, interface_name, interface_speed, verbose):
...
-    lane_len = len(breakout_file_input[interface_name]["lanes"].split(","))
 
-    if interface_name in breakout_file_input:
-        """ Check whether target breakout mode is available for the user-selected interface or  not"""
-        if target_brkout_mode not in breakout_file_input[interface_name]["breakout_modes"]:
-            click.secho('[ERROR] Target mode {} is not available for the port {}'. format(target_brkout_mode, interface_name), fg='red')
-            sys.exit(1)
+    # Get current breakout mode
+    cur_brkout_dict = config_db.get_table('BREAKOUT_CFG')
+    cur_brkout_mode = cur_brkout_mode = cur_brkout_dict[interface_name]["brkout_mode"]
```
Comment (li-pingmao) [inline config/main.py]: Should it be "cur_brkout_mode = cur_brkout_dict[interface_name]["brkout_mode"]"?

File: config/config_mgmt.py (hunk 1 above)
Comment (praveen-li) [reply config/config_mgmt.py:81]: Else is do nothing, in this case. So I left else case.

File: config/config_mgmt.py (hunk 6 above)
Comment (praveen-li) [reply config/config_mgmt.py]: I changed the way, It runs tests now.

File: config/config_mgmt.py (hunk 4 above)
Comment (praveen-li) [reply config/config_mgmt.py]: portJson contains only PORT Table Config.

File: config/config_mgmt.py (hunk 2 above)
Comment (praveen-li) [reply config/config_mgmt.py]: Changed the Test Code.

File: config/config_mgmt.py (hunk 3 above)
Comment (praveen-li) [reply config/config_mgmt.py]: Changed the Test Code.

**Review by praveen-li — COMMENTED** [review]
Kindly Merge this, we need these fixes in Sonic_Util, which was found during DPB debugging.

**Review by li-pingmao — APPROVED** [review]

**Pull request merged**